sdk/core/Azure.Core/src/Shared/AzureSasCredentialSynchronousPolicy.cs

```diff
@@ -29,13 +29,33 @@ public override void OnSendingRequest(HttpMessage message)
             {
                 signature = signature.Substring(1);
             }
+
             if (!query.Contains(signature))
             {
-                query = string.IsNullOrEmpty(query) ? '?' + signature : query + '&' + signature;
+                // Get the signature history so we know which string to replace in the query if it was previously set.
+                if (message.TryGetProperty(typeof(AzureSasSignatureHistory), out object? previousSignature) &&
+                    previousSignature != null &&
+                    query.Contains((string)previousSignature!))
+                {
+                    query = query.Replace((string)previousSignature!, signature);
+                }
+                else
```

Comment:
It always makes me nervous when there is an else case for an if-statement with a lot of clauses on it.  For example, if there was a previous signature stored on the message, but it was null, is this really the first time we're setting the signature?  Is it possible there would be a previous signature on the message but it isn't in the query?  Why would that happen?  Does this else case address all of the different reasons we could reach it?

Reply:
> if there was a previous signature stored on the message, but it was null, is this really the first time we're setting the signature?

If it was null, it doesn't matter. We effectively don't have a signature or a valid previous signature and we'll need to add one.

> Is it possible there would be a previous signature on the message but it isn't in the query? Why would that happen?

Yes, this is the scenario that describes the bug being fixed. 

1. The message is sent to the pipeline
2. `AzureSasCredentialSynchronousPolicy` adds the signature the query and to the previous signature property
3. The message is sent to the transport
4. The SAS credential is updated
5. The transport returns the response but it is a retriable failure
6. The message traverses the pipeline policies and `AzureSasCredentialSynchronousPolicy` evaluates the signature but it no longer matches, but the query does match the previous signature.
7. `AzureSasCredentialSynchronousPolicy` knows that the stale signature needs to be replaced rather than adding an additional signature with the new SAS value.

> Does this else case address all of the different reasons we could reach it?

I think so, but if you can think of an edge case not handled, please do!

Reply:
I see, ok, thank you for that explanation, it is incredibly helpful!

If that is the case (and since now that `AzureSasSignatureHistory`, this value can only be set in the policy), then do you even need to check for null or `Contains`?  You know the value isn't null since you're the only one setting it.  And I believe `string.Replace()` semantics are such that if the string isn't contained, it won't modify the value, so you would avoid doing two linear string checks which would be better for perf.

Would it make sense to take those out then and just do `previousSignature! as string`?  (I haven't tried it, maybe it doesn't work)

Reply:
We still need the `Contains` because if it is true, we need to do the replace. But if it is false, we need to add the query in the else block.

Reply:
That makes sense.  Curious: how would you end up with a value for the previous signature without it being appended to the query?

Reply:
It shouldn't happen normally, but if some other policy were to mess with the query, this could happen.